### ISSUE TITLE
Memory leak when writing pdf

### ISSUE THREAD

**Issue by ZuperZee** [Kozea/WeasyPrint#1977]
Calling `HTML(string=html_str).write_pdf("filename")` doesn't release the memory used.

Example (tracemalloc)
```py
import tracemalloc
from weasyprint import HTML

tracemalloc.start()


def write_pdf(html_str):
    HTML(string=html_str).write_pdf("./a.pdf")


html_str = f"<div>{'<div>Hello</div>' * 20}</div>"

write_pdf(html_str)
snap = tracemalloc.take_snapshot()
write_pdf(html_str)
stats = tracemalloc.take_snapshot().compare_to(snap, "traceback")

for stat in stats[:20]:
    print(stat)
```
Output
```txt
/home/zee/dev/weasyprint-mem-leak/venv/lib/python3.11/site-packages/cffi/api.py:335: size=743 KiB (+371 KiB), count=2 (+1), average=371 KiB
/usr/lib/python3.11/tracemalloc.py:558: size=69.5 KiB (+69.4 KiB), count=1270 (+1269), average=56 B
/home/zee/dev/weasyprint-mem-leak/venv/lib/python3.11/site-packages/weasyprint/css/__init__.py:740: size=139 KiB (+69.4 KiB), count=68 (+34), average=2089 B
/home/zee/dev/weasyprint-mem-leak/venv/lib/python3.11/site-packages/weasyprint/css/__init__.py:630: size=119 KiB (+59.4 KiB), count=80 (+40), average=1520 B
/home/zee/dev/weasyprint-mem-leak/venv/lib/python3.11/site-packages/weasyprint/formatting_structure/boxes.py:110: size=102 KiB (+51.2 KiB), count=254 (+127), average=413 B
/home/zee/dev/weasyprint-mem-leak/venv/lib/python3.11/site-packages/weasyprint/css/__init__.py:647: size=70.3 KiB (+35.2 KiB), count=540 (+270), average=133 B
/home/zee/dev/weasyprint-mem-leak/venv/lib/python3.11/site-packages/fontTools/ttLib/tables/_h_m_t_x.py:64: size=109 KiB (+27.2 KiB), count=1995 (+498), average=56 B
/home/zee/dev/weasyprint-mem-leak/venv/lib/python3.11/site-packages/weasyprint/css/__init__.py:692: size=37.4 KiB (+18.7 KiB), count=128 (+64), average=299 B
/home/zee/dev/weasyprint-mem-leak/venv/lib/python3.11/site-packages/weasyprint/layout/percent.py:137: size=36.5 KiB (+18.2 KiB), count=512 (+256), average=73 B
/home/zee/dev/weasyprint-mem-leak/venv/lib/python3.11/site-packages/fontTools/ttLib/tables/O_S_2f_2.py:575: size=0 B (-17.7 KiB), count=0 (-324)
/home/zee/dev/weasyprint-mem-leak/venv/lib/python3.11/site-packages/weasyprint/css/__init__.py:93: size=25.6 KiB (+13.2 KiB), count=377 (+192), average=69 B
/home/zee/dev/weasyprint-mem-leak/venv/lib/python3.11/site-packages/weasyprint/css/__init__.py:616: size=26.2 KiB (+13.1 KiB), count=160 (+80), average=168 B
/home/zee/dev/weasyprint-mem-leak/venv/lib/python3.11/site-packages/weasyprint/css/computed_values.py:407: size=24.2 KiB (+12.1 KiB), count=376 (+188), average=66 B
/home/zee/dev/weasyprint-mem-leak/venv/lib/python3.11/site-packages/weasyprint/layout/percent.py:97: size=21.3 KiB (+10.7 KiB), count=328 (+164), average=66 B
/home/zee/dev/weasyprint-mem-leak/venv/lib/python3.11/site-packages/weasyprint/css/__init__.py:162: size=18.9 KiB (+9664 B), count=182 (+91), average=106 B
/home/zee/dev/weasyprint-mem-leak/venv/lib/python3.11/site-packages/weasyprint/formatting_structure/boxes.py:360: size=18.6 KiB (+9504 B), count=88 (+44), average=216 B
/home/zee/dev/weasyprint-mem-leak/venv/lib/python3.11/site-packages/fontTools/subset/__init__.py:2761: size=16.4 KiB (+8384 B), count=314 (+157), average=53 B
/home/zee/dev/weasyprint-mem-leak/venv/lib/python3.11/site-packages/weasyprint/css/__init__.py:90: size=16.3 KiB (+8344 B), count=298 (+149), average=56 B
/home/zee/dev/weasyprint-mem-leak/venv/lib/python3.11/site-packages/weasyprint/css/__init__.py:751: size=15.5 KiB (+7920 B), count=180 (+90), average=88 B
/home/zee/dev/weasyprint-mem-leak/venv/lib/python3.11/site-packages/weasyprint/css/__init__.py:695: size=12.8 KiB (+6528 B), count=4 (+2), average=3264 B
```

Example (psutil)
```py
import psutil
import os
from weasyprint import HTML


def write_pdf(html_str):
    HTML(string=html_str).write_pdf("./a.pdf")


html_str = f"<div>{'<div>Hello</div>' * 20}</div>"

print(psutil.Process(os.getpid()).memory_info())
for i in range(20):
    write_pdf(html_str)
    print(psutil.Process(os.getpid()).memory_info())
```
Output
```txt
pmem(rss=46710784, vms=64737280, shared=13434880, text=3194880, lib=0, data=34611200, dirty=0)
pmem(rss=53391360, vms=69718016, shared=15953920, text=3194880, lib=0, data=39088128, dirty=0)
pmem(rss=54624256, vms=70316032, shared=16334848, text=3194880, lib=0, data=39305216, dirty=0)
pmem(rss=56201216, vms=72634368, shared=16715776, text=3194880, lib=0, data=41242624, dirty=0)
pmem(rss=56709120, vms=73224192, shared=16711680, text=3194880, lib=0, data=41832448, dirty=0)
pmem(rss=58286080, vms=74317824, shared=17092608, text=3194880, lib=0, data=42545152, dirty=0)
pmem(rss=59854848, vms=75587584, shared=17473536, text=3194880, lib=0, data=43433984, dirty=0)
pmem(rss=61444096, vms=77905920, shared=17854464, text=3194880, lib=0, data=45371392, dirty=0)
pmem(rss=63016960, vms=79183872, shared=18235392, text=3194880, lib=0, data=46268416, dirty=0)
pmem(rss=64577536, vms=80453632, shared=18616320, text=3194880, lib=0, data=47157248, dirty=0)
pmem(rss=62644224, vms=78520320, shared=15949824, text=3194880, lib=0, data=47890432, dirty=0)
pmem(rss=63537152, vms=79699968, shared=16330752, text=3194880, lib=0, data=48689152, dirty=0)
pmem(rss=64319488, vms=80461824, shared=16711680, text=3194880, lib=0, data=49070080, dirty=0)
pmem(rss=65081344, vms=81223680, shared=17092608, text=3194880, lib=0, data=49451008, dirty=0)
pmem(rss=65847296, vms=81985536, shared=17473536, text=3194880, lib=0, data=49831936, dirty=0)
pmem(rss=66633728, vms=82747392, shared=17854464, text=3194880, lib=0, data=50212864, dirty=0)
pmem(rss=67026944, vms=83124224, shared=17854464, text=3194880, lib=0, data=50589696, dirty=0)
pmem(rss=68059136, vms=83886080, shared=18235392, text=3194880, lib=0, data=50970624, dirty=0)
pmem(rss=69320704, vms=85843968, shared=18616320, text=3194880, lib=0, data=52547584, dirty=0)
pmem(rss=70275072, vms=86339584, shared=18997248, text=3194880, lib=0, data=52662272, dirty=0)
pmem(rss=71819264, vms=87605248, shared=19378176, text=3194880, lib=0, data=53547008, dirty=0)
```

WeasyPrint info `python3 -m weasyprint --info`
```
System: Linux
Machine: x86_64
Version: #1 SMP Fri Jan 27 02:56:13 UTC 2023
Release: 5.15.90.1-microsoft-standard-WSL2

WeasyPrint version: 60.1
Python version: 3.11.5
Pydyf version: 0.8.0
Pango version: 14407
```

**Comment by liZe** [comment]
Hi!

Thanks for your report.

Python’s memory management is opaque by design and can’t be managed manually (even using `del` or manually calling the garbage collector). The interpreter loves to keep allocated memory around, but it doesn’t mean that there’s a memory leak.

20 loops is not enough to find memory leaks. Here’s what I get with 5000 loops:
```
0 pmem(rss=65290240, vms=308670464, shared=25182208, text=4096, lib=0, data=66711552, dirty=0)
1 pmem(rss=70664192, vms=312553472, shared=25579520, text=4096, lib=0, data=71229440, dirty=0)
2 pmem(rss=75231232, vms=316203008, shared=25976832, text=4096, lib=0, data=74964992, dirty=0)
3 pmem(rss=77979648, vms=318853120, shared=25976832, text=4096, lib=0, data=77615104, dirty=0)
4 pmem(rss=79458304, vms=321310720, shared=26374144, text=4096, lib=0, data=79683584, dirty=0)
5 pmem(rss=83550208, vms=324947968, shared=26771456, text=4096, lib=0, data=82927616, dirty=0)
[…]
200 pmem(rss=172130304, vms=412708864, shared=26771456, text=4096, lib=0, data=170844160, dirty=0)
201 pmem(rss=172130304, vms=412708864, shared=26771456, text=4096, lib=0, data=170844160, dirty=0)
202 pmem(rss=172527616, vms=413106176, shared=27168768, text=4096, lib=0, data=170844160, dirty=0)
203 pmem(rss=172924928, vms=413503488, shared=27566080, text=4096, lib=0, data=170844160, dirty=0)
204 pmem(rss=173322240, vms=413503488, shared=27963392, text=4096, lib=0, data=170844160, dirty=0)
205 pmem(rss=173326336, vms=413900800, shared=27963392, text=4096, lib=0, data=170848256, dirty=0)
[…]
500 pmem(rss=172363776, vms=412987392, shared=26771456, text=4096, lib=0, data=171671552, dirty=0)
501 pmem(rss=172363776, vms=412987392, shared=26771456, text=4096, lib=0, data=171671552, dirty=0)
502 pmem(rss=172761088, vms=413384704, shared=27168768, text=4096, lib=0, data=171671552, dirty=0)
503 pmem(rss=173158400, vms=413782016, shared=27566080, text=4096, lib=0, data=171671552, dirty=0)
504 pmem(rss=173555712, vms=413782016, shared=27963392, text=4096, lib=0, data=171671552, dirty=0)
505 pmem(rss=173555712, vms=414179328, shared=27963392, text=4096, lib=0, data=171671552, dirty=0)
[…]
800 pmem(rss=172388352, vms=412905472, shared=26771456, text=4096, lib=0, data=171614208, dirty=0)
801 pmem(rss=172388352, vms=412905472, shared=26771456, text=4096, lib=0, data=171614208, dirty=0)
802 pmem(rss=172785664, vms=413302784, shared=27168768, text=4096, lib=0, data=171614208, dirty=0)
803 pmem(rss=173182976, vms=413700096, shared=27566080, text=4096, lib=0, data=171614208, dirty=0)
804 pmem(rss=173580288, vms=413700096, shared=27963392, text=4096, lib=0, data=171614208, dirty=0)
805 pmem(rss=173580288, vms=414097408, shared=27963392, text=4096, lib=0, data=171614208, dirty=0)
[…]
994 pmem(rss=171601920, vms=412217344, shared=25579520, text=4096, lib=0, data=172118016, dirty=0)
995 pmem(rss=171999232, vms=412217344, shared=25976832, text=4096, lib=0, data=172118016, dirty=0)
996 pmem(rss=171999232, vms=412614656, shared=25976832, text=4096, lib=0, data=172118016, dirty=0)
997 pmem(rss=172396544, vms=412614656, shared=26374144, text=4096, lib=0, data=172118016, dirty=0)
998 pmem(rss=172396544, vms=413011968, shared=26374144, text=4096, lib=0, data=172118016, dirty=0)
999 pmem(rss=172396544, vms=413011968, shared=26374144, text=4096, lib=0, data=172118016, dirty=0)
[…]
4995 pmem(rss=172474368, vms=412962816, shared=25735168, text=4096, lib=0, data=172879872, dirty=0)
4996 pmem(rss=172474368, vms=413360128, shared=25735168, text=4096, lib=0, data=172879872, dirty=0)
4997 pmem(rss=172871680, vms=413360128, shared=26132480, text=4096, lib=0, data=172879872, dirty=0)
4998 pmem(rss=172871680, vms=413757440, shared=26132480, text=4096, lib=0, data=172879872, dirty=0)
4999 pmem(rss=173268992, vms=413757440, shared=26529792, text=4096, lib=0, data=172879872, dirty=0)
```

After ~150 loops, the memory use is quite stable, so there’s no obvious memory leak for your example as far as I can tell.

**Comment by ZuperZee** [comment]
Ah. My bad. Thanks for clarifying.

**Comment by ralphie0112358** [comment]
For anyone else that ends up here...  I found `html.write_pdf()` tends to hang onto memory as observed here, likely it would EVENTUALLY be freed by gc, but python and the os can be greedy and never release the memory fully.  I prefer to not consume memory like this if possible.  In my loop testing I could attribute a majority of the memory to the `ffi.unpack()` call which loads the fonts (I believe).
https://github.com/Kozea/WeasyPrint/blob/4d93769a06e2a5a9d158daf2b825bbcff143ce4f/weasyprint/pdf/stream.py#L27

If I cache the loaded fonts then the fonts need only be loaded once and if your service is mostly reusing the same fonts, I believe you will save some memory usage.  I was able to cache the fonts using something like:
```
class PdfWrite:
    _fonts = {}
    def write_pdf(self, html_str):
        doc = HTML(string=html_str).render()
        doc.fonts = self._fonts
        doc.write_pdf("./a.pdf")

PdfWrite().write_pdf(html_str)
```

Of course, now you have a cache that will never free unless you manually clear it `PdfWrite._fonts = {}`, then gc should do its thing.  

I performed a similar test as above with before and after the caching strategy applied

Original output in my environment:
```
pmem(rss=56197120, vms=77615104, shared=17039360, text=4096, lib=0, data=39755776, dirty=0)
pmem(rss=63143936, vms=233340928, shared=19660800, text=4096, lib=0, data=61304832, dirty=0)
pmem(rss=64311296, vms=234008576, shared=20054016, text=4096, lib=0, data=61591552, dirty=0)
pmem(rss=65351680, vms=234909696, shared=20082688, text=4096, lib=0, data=62492672, dirty=0)
pmem(rss=66355200, vms=236978176, shared=20344832, text=4096, lib=0, data=64180224, dirty=0)
pmem(rss=67891200, vms=238026752, shared=20738048, text=4096, lib=0, data=64847872, dirty=0)
pmem(rss=69349376, vms=239448064, shared=21000192, text=4096, lib=0, data=65888256, dirty=0)
pmem(rss=70955008, vms=241655808, shared=21393408, text=4096, lib=0, data=67715072, dirty=0)
pmem(rss=69537792, vms=240119808, shared=19595264, text=4096, lib=0, data=68083712, dirty=0)
pmem(rss=70586368, vms=241451008, shared=19857408, text=4096, lib=0, data=69033984, dirty=0)
pmem(rss=71372800, vms=242208768, shared=20250624, text=4096, lib=0, data=69410816, dirty=0)
pmem(rss=72159232, vms=259756032, shared=20643840, text=4096, lib=0, data=86568960, dirty=0)
pmem(rss=72945664, vms=260517888, shared=21037056, text=4096, lib=0, data=86949888, dirty=0)
pmem(rss=73748480, vms=260833280, shared=21430272, text=4096, lib=0, data=86884352, dirty=0)
pmem(rss=75325440, vms=262094848, shared=21823488, text=4096, lib=0, data=87764992, dirty=0)
pmem(rss=76423168, vms=263000064, shared=21905408, text=4096, lib=0, data=88670208, dirty=0)
pmem(rss=77201408, vms=264880128, shared=22167552, text=4096, lib=0, data=90169344, dirty=0)
pmem(rss=78950400, vms=266117120, shared=22560768, text=4096, lib=0, data=91025408, dirty=0)
pmem(rss=80687104, vms=267399168, shared=23085056, text=4096, lib=0, data=91926528, dirty=0)
pmem(rss=82284544, vms=268668928, shared=23478272, text=4096, lib=0, data=92815360, dirty=0)
pmem(rss=82657280, vms=270290944, shared=23367680, text=4096, lib=0, data=94437376, dirty=0)
```


Apply font cache
```
import psutil
import os
from weasyprint import HTML


class PdfWriter:
    _fonts = {}

    def write_pdf(self, html_str):
        doc = HTML(string=html_str).render()
        doc.fonts = self._fonts
        doc.write_pdf("./a.pdf")


html_str = f"<div>{'<div>Hello</div>' * 20}</div>"

print(psutil.Process(os.getpid()).memory_info())
for i in range(20):
    PdfWriter().write_pdf(html_str)
    print(psutil.Process(os.getpid()).memory_info())
```
Output
```
pmem(rss=56012800, vms=77611008, shared=17039360, text=4096, lib=0, data=39751680, dirty=0)
pmem(rss=62828544, vms=166236160, shared=19398656, text=4096, lib=0, data=61161472, dirty=0)
pmem(rss=63090688, vms=233725952, shared=19529728, text=4096, lib=0, data=61296640, dirty=0)
pmem(rss=63614976, vms=234582016, shared=19791872, text=4096, lib=0, data=61771776, dirty=0)
pmem(rss=64139264, vms=235474944, shared=19922944, text=4096, lib=0, data=62283776, dirty=0)
pmem(rss=64925696, vms=236367872, shared=20185088, text=4096, lib=0, data=62795776, dirty=0)
pmem(rss=65581056, vms=237264896, shared=20316160, text=4096, lib=0, data=63311872, dirty=0)
pmem(rss=66498560, vms=238153728, shared=20578304, text=4096, lib=0, data=63819776, dirty=0)
pmem(rss=65667072, vms=236249088, shared=19615744, text=4096, lib=0, data=63819776, dirty=0)
pmem(rss=65798144, vms=236630016, shared=19746816, text=4096, lib=0, data=63819776, dirty=0)
pmem(rss=66060288, vms=237010944, shared=20008960, text=4096, lib=0, data=63819776, dirty=0)
pmem(rss=66191360, vms=237391872, shared=20140032, text=4096, lib=0, data=63819776, dirty=0)
pmem(rss=66322432, vms=237772800, shared=20271104, text=4096, lib=0, data=63819776, dirty=0)
pmem(rss=66715648, vms=238153728, shared=20533248, text=4096, lib=0, data=63819776, dirty=0)
pmem(rss=65937408, vms=236249088, shared=19623936, text=4096, lib=0, data=63819776, dirty=0)
pmem(rss=66068480, vms=236630016, shared=19755008, text=4096, lib=0, data=63819776, dirty=0)
pmem(rss=66330624, vms=237010944, shared=20017152, text=4096, lib=0, data=63819776, dirty=0)
pmem(rss=66461696, vms=237391872, shared=20148224, text=4096, lib=0, data=63819776, dirty=0)
pmem(rss=66723840, vms=237772800, shared=20410368, text=4096, lib=0, data=63819776, dirty=0)
pmem(rss=66854912, vms=313655296, shared=20541440, text=4096, lib=0, data=72343552, dirty=0)
pmem(rss=66560000, vms=313466880, shared=19853312, text=4096, lib=0, data=73678848, dirty=0)
```

I appreciate any feedback from other users or core weasyprint devs on this approach or if there might be another solution or fix in the future.  Thank you.

**Comment by liZe** [comment]
> I appreciate any feedback from other users or core weasyprint devs on this approach or if there might be another solution or fix in the future. Thank you.

Thanks a lot, it may be useful for other users! 💜

We’ve tried to play with Python’s garbage collector many times, but the different "solutions" we’ve tested so far depend a lot on the version of Python, the OS, the quantity of available RAM, the use case, etc. For now, we let Python manage memory the way it wants, and only fix the real memory leaks. But users can of course tweak the memory depending on their use case, so thanks for your snippet!

**Comment by liZe** [comment]
Hello everyone!

This issue should be fixed with #2130.
